### DISCUSSION TITLE
How to get a fresh instance of code-server to automatically add the location of a programming language location to PATH on startup?

### DISCUSSION THREAD

**Discussion by RodogInfinite** [coder/code-server#4734] [Q&A] [answered]
I'm currently using a container to install Rust and then install code-server automatically with Debian as the base OS. I have a run script in the container that can echo the environment variable where Rust can be found, `echo $RUST_ENV`, before starting up code-server. I'm finding that code-server has its own `$PATH` separate from the container itself and, therefore, it can't recognize where Rust is after a fresh rebuild; however, I am able to access `$RUST_ENV` and manually add it to the code-server `$PATH` using `export $RUST_ENV` from within a code-server terminal when running. Is there a way to get `$RUST_ENV` added to the code-server's `$PATH` variable on startup after a fresh build without having to do this manually every time? I often don't need to make changes to the container for stretches of time so I forget this step and it takes time for me to figure it out again which could be spent on other things.

[Edit 1] What I'm envisioning is a flag that doesn't exist currently. Something like `code-server --lang_path /path/to/your/language`, but if another method that achieves the same results exists, then I'd be happy to implement that so long as it can be automated in the container.

[Edit 2] After more digging it seems like these issues are related: [Terminal does not load user paths or env variables](https://github.com/coder/code-server/issues/226) and [https://github.com/testcab/ansible-role-code-server/issues/9](https://github.com/testcab/ansible-role-code-server/issues/9)

**Comment by jsjoeio** [comment]
I ran into similar issues using code-server in Coder with Rust. While the paths won't be exactly the same, maybe it will give a clue on how you can solve your problem?

https://github.com/jsjoeio/joe-image/blob/main/Dockerfile#L76-L81

**Reply by RodogInfinite** [reply]
This gives me some idea. I'm currently using [Singularity](https://sylabs.io/singularity) as a container platform and it's more isolated from the host system than Docker so it may not translate directly. Also, I'm assuming you have coder already installed on your host system before using building from that Dockerfile since I'm not seeing it being installed within? I'm installing code-server within the container so it the image can be handed off with nearly everything it needs to run as long as Singularity is installed

**Reply by jsjoeio** [reply]
> I'm assuming you have coder already installed on your host system before using building from that Dockerfile since I'm not seeing it being installed within?

Yes, exactly! This runs in a workspace on our product which then works with code-server. 

I hope you can get a solution that works with Singularity 🤞

**Comment by RodogInfinite** [comment, marked as answer]
I had to put this down for a long while. I have a solution, but I'm not sure if it's the best one. So my Singularity container installs code-server in the `%post` section (executes during building) and then in the `%run` section (executes when the container is launched) I have:

```
%run
# Check the settings.json to see if it's empty
settings_file="/root/.local/share/code-server/User/settings.json"
if ! [ -s "$settings_file" ]; then
    # File is empty; therefore, jq append below will not work. Create empty JSON object '{}'
    echo '{}' > "$settings_file"
fi
# Modify code-server's settings.json to include the path to Rust
echo "$( jq '."terminal.integrated.env.linux"={"PATH":"/usr/lib/code-server/vendor/modules/code-oss-dev/bin:/opt/rust/cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"}' /root/.local/share/code-server/User/settings.json )" > /root/.local/share/code-server/User/settings.json

# Start code-server
code-server
```
What it's doing is adding the path string I define into the settings.json via the `terminal.integrated.env.linux` key using [jq](https://stedolan.github.io/jq/manual/) to modify the file. Afterwards my settings.json looks like:
```
{
  "terminal.integrated.env.linux": {
    "PATH": "/usr/lib/code-server/vendor/modules/code-oss-dev/bin:/opt/rust/cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
  }
}
```
Now when I run `cargo` it just works. I hope this helps someone in the future. If anyone finds a better way please let us know!
